### PR TITLE
Add readdir()

### PR DESCRIPTION
**Pull request by mattn**
glob() is bit difficult to use to get file list in Vim script since it depend on suffixes or wildignore. So this change add new API `readdir()`.

readdir() take 1 or 2 arguments. first one is a path to the directory to be listed files. sencond one is expr to add or ignore for the result. 1 to add the filename. 0 to ignore. If you set -1 to expr, readdir() stop to read. Below is test code.

```vim
func Test_readdir()
  call mkdir('Xdir')
  call writefile([], 'Xdir/foo.txt')
  call writefile([], 'Xdir/bar.txt')
  call mkdir('Xdir/dir')

  let files = readdir('Xdir')
  call assert_equal(['bar.txt', 'dir', 'foo.txt'], files)

  let files = readdir('Xdir', {x->stridx(x,'f')!=-1})
  call assert_equal(['foo.txt'], files)

  let l = []
  let files = readdir('Xdir', {x->len(add(l, x)) == 2 ? -1 : 1})
  call assert_equal(1, len(files))

  call delete('Xdir', 'rf')
endfunc
```

### CODE REVIEW

File: runtime/doc/eval.txt [hunk 1]
```diff
@@ -2282,6 +2282,9 @@ py3eval({expr})			any	evaluate |python3| expression
 pyxeval({expr})			any	evaluate |python_x| expression
 range({expr} [, {max} [, {stride}]])
 				List	items from {expr} to {max}
+readdir({directory} [, {max}])
```
Comment (k-takata) [inline runtime/doc/eval.txt]: `{max}` -> `{expr}`

**Comment by chrisbra** [comment]
it might be silly, but would it make sense to have an indication on the returned list, whether the resulting entry is a directory?
E.g. instead of this:
```viml
let files = readdir('Xdir')
call assert_equal(['bar.txt', 'dir', 'foo.txt'], files)
```
```viml
let files = readdir('Xdir')
call assert_equal(['bar.txt', 'dir/', 'foo.txt'], files)
```
one more thing to think about it, would it make sense to return a recursive list of matches?

**Comment by mattn** [comment]
I'm thinking that return recursive list is over work. But it can be done with:
```vim
function! s:tree(dir)
  return {a:dir : map(readdir(a:dir), {_,x -> isdirectory(x) ? {x : s:tree(a:dir . '/' . x)} : x})}
endfunction

echo s:tree(".")
```
Appending suffix '/' may be useful to know the item is directory easily.

**Comment by chrisbra** [comment]
that example should be added to the documentation :)

File: runtime/doc/eval.txt [hunk 2]
```diff
@@ -6414,7 +6414,13 @@ readdir({directory} [, {expr}])
...
 		evaluate {expr}. If {expr} return -1, readdir stop to read.
 		If it return 0, result does not contain the file name.
-		If return 1, result contain the file name.
+		If return 1, result contain the file name. If you want to get
+		directory tree: >
+                  function! s:tree(dir)
+                      return {a:dir : map(readdir(a:dir), {_,x ->
+                      \ isdirectory(x) ? {x : s:tree(a:dir . '/' . x)} : x})}
+                  endfunction
+                  echo s:tree(".")
```
Comment (k-takata) [inline runtime/doc/eval.txt]: `<` is needed in the top of this line.

**Comment by Shougo** [comment]
https://github.com/vim/vim/issues/696
https://github.com/vim/vim/pull/1099

It may fix the `delete('rf')`problems.
`delete('rf')` uses `glob()` internally.
But it should use `readdir()` instead.

**Comment by brammool** [comment]
Yasuhiro Matsumoto wrote:

> glob() is bit difficult to use to get file list in Vim script since it
> depend on suffixes or wildignore. So this change add new API
> `readdir()`.
> 
> readdir() take 1 or 2 arguments. first one is a path to the directory
> to be listed files. sencond one is expr to add or ignore for the
> result. 1 to add the filename. 0 to ignore. If you set -1 to expr,
> readdir() stop to read. Below is test code.

Hmm, this results in a large overlap of functionality between glob() and
readdir().  Each covers a part of the functionality, a lot can be done
with either, and there will still be things that neither can do and will
have to be added later.

Using readdir() for recursive stuff is a bit complicated, while with
glob() that's easy.  The use of a filter function is nice for readdir()
and lacking from glob().

Also, glob() has gained several optional arguments over the year, which
now make it clumsy to use.

How about this instead:
	glob({expr} [, {nosuf} [, {list} [, {alllinks}]]])
	glob({expr}, {dict})

The first form is what we already have, we keep supporting it.
The second form passes arguments in a dictionary.  Members are:
	nosuf		as the {nosuf} argument
	list		as the {list} argument (could be default on)
	alllinks	as the {alllinks} argument
	filter		function like the readdir() function has

Then you can use:
	global('*', filter)
like the proposed readdir() function.
But you can also do:
	global('**', filter)
To make it work recursively.

We can also add an option to add a slash to directory names:
	dirslash

>   let files = readdir('Xdir', {x->stridx(x,'f')!=-1})

If you write a lambda function, please put spaces around the "->",
otherwise it starts looking like C code :-).


-- 
hundred-and-one symptoms of being an internet addict:
93. New mail alarm on your palmtop annoys other churchgoers.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Pull request closed without merging**